### ISSUE TITLE
Bug: [no-unnecessary-condition] warns about conditional assignment to index signature

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [typescript-eslint/typescript-eslint#6632]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=4.9.5&sourceType=module&code=MYewdgzgLgBAZiEAuGAlApqATgEwDzRYCWYA5gDQxgCuAtgEbpYB8MAvDAN4C+A3ALAAoUJFj0AhlhSESpdjADkAa3QBPBQMFCEIANoSsAXRgB+ExwAMvGAHob8RPsnGiEKiFgkiUIuIA2RABe6DiU9NSwIjjeROD+MOIQEESkYLToYJ5ucH7ipKQhCW7UYGCY6EmSqgB02o4GxgDUHACMmkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```typescript
const foo: Record<string, number> = {};
const bar: string = 'key';

foo[bar] ??= 0; // foo[bar] is not initialized, but conditional assignment is flagged as unnecessary.
foo[bar] += 1;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No warning.

### Actual Result

Unnecessary conditional, expected left-hand side of `??` operator to be possibly null or undefined. 4:1 - 4:9

### Additional Info

This is basically https://github.com/typescript-eslint/typescript-eslint/issues/6264 which was fixed in https://github.com/typescript-eslint/typescript-eslint/pull/6452; I guess it resurfaced because the rule now checks logical assignments as of https://github.com/typescript-eslint/typescript-eslint/pull/6594.

**Comment by bradzacher** [comment]
This is working as intended.
According to the types - the type of `foo[bar]` is `string`. Additionally the assignment syntax `foo[bar] ??= 0` is equivalent to `foo[bar] ?? (foo[bar] = 0)`

So combining the two - the rule will report on `foo[bar] ?? thing` because the types say that the left side of the `??` is never nullish. Which logically follows that the rule should also report on the use of `??=` for this case because according to the types it's unnecessary.

This is a bit of a weakness of TS record types - it doesn't encode `undefined`.

Now the interesting thing is that there is the `noUncheckedIndexAccess` compiler option that will union `undefined` into the index type at read time to enforce correct code without creating a hole at write time like manually unioning `undefined` would do.

This option works as expected with the expanded form of the assignment operator, however it doesn't work with the assignment operator - because even though the assignment is both a read and a write, TS doesn't treat this as a read location and thus doesn't add `undefined` to the type we see.

[playground](https://typescript-eslint.io/play#ts=4.9.5&sourceType=module&code=MYewdgzgLgBAZiEAuGAlApqATgEwDzRYCWYA5gDQxgCuAtgEbpYB8MAvDAN4C+A3ALAAoUJFj0AhlhSESpdjADkAa3QBPBQMFCEIANoSsAXRgB+EzAAUO-ZOMcADAEpN1g8bMPNQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAqkWgALdNADW6ACYBJIjPQAPWQEFo0dCTKUO6XgF8QhoA)

So what is the correct solution for this? TBH I don't have a great one right now. The assignment operator change was omly just added in https://github.com/typescript-eslint/typescript-eslint/pull/6594. Perhaps it's something we need to add an option for people to opt out of? Perhaps there's a way we could manually code the heuristics if the compiler option is on? Perhaps this is something we could chat to the TS team about correcting?
I'm not sure exactly.

Cc @typescript-eslint/triage-team for opinions

**Comment by ehoogeveen-medweb** [comment]
I see, that's an unfortunate difference in behavior.

As I've expressed elsewhere, I don't think `noUncheckedIndexedAccess` is a workable compiler option for most codebases because of how difficult it becomes to convince TypeScript that "yes, this access is safe" without sprinkling non-null assertions around everywhere (I've [experimented](https://github.com/microsoft/TypeScript/pull/39560#issuecomment-1223915079) with workarounds in my code but they were clunky and introduced additional unsafe edge cases so I ended up removing them again).

I personally think indexed access signatures should always be considered optional by `typescript-eslint` even if TypeScript itself doesn't represent them this way. I don't expect the linter to warn me about unsafe access (although I would be very interested in trying a rule that tries to emulate `noUncheckedIndexedAccess` without the drawbacks), but I _would_ like it to avoid warning about things that aren't strictly true.

But I'm sure that's easier said than done. I think what I'll try to do as a workaround is to wrap more of my records in `Partial<>`.

**Comment by Josh-Cena** [comment]
I'm okay with aligning the behavior with https://github.com/typescript-eslint/typescript-eslint/pull/6452 (i.e. special-casing index signature types). Although I do recommend you turn on `noUncheckedIndexAccess`.

**Comment by jtbandes** [comment]
I ran into this too, with `noUncheckedIndexedAccess: true`. For example: [playground link](https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MYewdgzgLgBNCuAzRAuGAlApqATgEwB5ocBLMAcwBoYx4BbAI0xwD4YBeGAbwF8BuAFADQkWADcAhgBs0tRsw5woSRAG0ARAAsS6gLqCEyDdr0wA-Gc4AmQUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2RGwAqkWgALdNADW6ACYBJIjPQAPWQEFo0dCTKUO6XgF8QhoA)


```ts
const stuff: Record<string, number> = {};

const val: number = stuff["hi"];  // TS error, noUncheckedIndexedAccess:true means stuff["hi"] is number|undefined
stuff["hi"] ??= 2; // no-unnecessary-condition error, eslint thinks stuff["hi"] cannot be undefined
```

**Comment by Josh-Cena** [comment]
This is weird, but it's expected. If `stuff.hi` includes `undefined` in a write position, then you could legally write `stuff.hi = undefined`, but that breaks the `Record<string, number>` contract, so the property type in a write position only includes `number`. I guess we should just disable the check for index signatures.

**Comment by jtbandes** [comment]
> that breaks the `Record<string, number>` contract

What are you saying the contract is here? In my mind, if I set `noUncheckedIndexedAccess: true`, the contract is that `stuff.hi` is either a number or undefined, so the contract wouldn't be broken. Maybe the issue would be with `Object.entries(stuff)` now containing some undefined values?

**Comment by Josh-Cena** [comment]
The contract is, if you read `stuff.hi`, the return value may be `undefined`, because the key may not exist; however, if you check `"hi" in stuff`, then `stuff.hi` can only be `number`, because there's no chance that a property exists but has value `undefined`. This is why you can't write `undefined` to `stuff.hi`. `Record<string, number>` is still different from `Record<string, number | undefined>`

**Comment by jtbandes** [comment]
> This is why you can't write `undefined` to `stuff.hi`.

Makes sense. However, in `stuff["hi"] ??= 2;`, I'm not trying to do so.

**Comment by Josh-Cena** [comment]
The point is, if the type of `stuff.hi` includes `undefined` in a write position, then `stuff.hi = undefined` would be legal. The type of a symbol must first be resolved before checking the validity of the assignment to it.

**Comment by jtbandes** [comment]
Since `stuff["hi"] ??= 2` effectively means `stuff["hi"] = (stuff["hi"] ?? 2)`, I was thinking of that as two separate operations (read & subsequent write) where the read source & write target can have different types. The read would return `number|undefined`, but the write would only accept `number` (so e.g. the right hand side of `??` wouldn't be allowed to be nullable).

**Comment by Josh-Cena** [comment]
No, `stuff.hi ??= 2` is equivalent to `stuff.hi ?? (stuff.hi = 2)`.

**Comment by jtbandes** [comment]
> `stuff.hi ?? (stuff.hi = 2)`

Sorry to continue beating an apparently dead horse, but it still seems to me that this is two operations, the first being a read (which returns `number | undefined`), and the second being a write which only accepts `number`. That is, the lhs is evaluated first, returning `number | undefined`, and if it's not undefined then the assignment in the rhs happens (which is just an assignment of a number literal to a number property).

But If the TS APIs don't make it easy to model it as two operations this way then I guess that is an issue.

**Comment by Josh-Cena** [comment]
Yeah, the way the rule is written is that we check the expression `stuff.hi ??= 2` and then we query the TS compiler API to tell us what the type of the LHS is, and TS says "`string`", so we are convinced it cannot be `undefined`. TS does not tell us what happens if it's hypothetically in a read position.

**Comment by bradzacher** [comment]
Filed a bug with TS - https://github.com/microsoft/TypeScript/issues/53303
IMO this should be fixed in TS itself cos even TS behaves differently with assignment operators vs their expanded form. But we'll see what they say.

If not, then we either must implement our own logic to determine if the LHS might be an unchecked index access or not, or we might just disable checks on `??=`. First port of call is the types, then we can make a decision about next steps.

**Comment by karlhorky** [comment]
I agree from a logical perspective that TypeScript itself should provide this information.

But since [it seems like they won't](https://github.com/microsoft/TypeScript/issues/53303#issuecomment-1474350135), what do you think about the following unconventional option?

Instead of [dropping support for checking for unnecessary logical assignment operators](https://github.com/microsoft/TypeScript/issues/53303#issuecomment-1474485817), what about instead leaning in and:

1. Making `@typescript-eslint/` packages a new type of TypeScript client, which reads the `tsconfig.json` for a handful of simple features (to be clear, not implementing much differently, only if it absolutely makes sense). Here the `compilerOptions.noUncheckedIndexedAccess` setting could be read to extend types in this assignment position.
2. Special-casing the `??=` operator with this new data.

This would also open the door to implement other features based on user configuration, where TypeScript does not give the expected types - of course each new feature here would need to be chosen carefully, because of potentially opening up incompatibilities with TypeScript.

**Comment by karlhorky** [comment]
@bradzacher @Josh-Cena as an alternative what do you think about publishing a new version either rolling back the change to the `@typescript-eslint/no-unnecessary-condition` rule with `??` and `??=` operators?

Every time Renovate updates all of our projects, it's breaking our linting and we need to manually revert to the old version `5.54.1` :(

**Comment by bradzacher** [comment]
Regardless of TS's position - we'll need the special case for backwards-compatibility.

If @Josh-Cena or anyone else has the time to special case it before the next release, then we can special case it.
Otherwise we will probably need to roll-back the change and then re-release it later with the special case.

**Comment by Josh-Cena** [comment]
I'll try to put together a PR today

**Comment by karlhorky** [comment]
@Josh-Cena @bradzacher any news? I see #6762 is open now, do you think it needs much more work?

**Comment by karlhorky** [comment]
It seems like the TypeScript PR https://github.com/microsoft/TypeScript/pull/54777 was merged, which closed the issue in the TypeScript repo:

- https://github.com/microsoft/TypeScript/issues/53305#event-10100428639

Will this mean that this issue will "fix itself" once the version of TypeScript (eg. maybe 5.2) is released containing the changes from the TypeScript PR?

**Comment by Peeja** [comment]
Copying [the answer](https://github.com/typescript-eslint/typescript-eslint/pull/6762#issuecomment-1680765475) from @Josh-Cena here for completeness: 😀 

> That PR changes the quick info of the write type so it excludes undefined (and the programmatic API always excludes it), which makes this PR even more necessary :)

**Comment by ehoogeveen-medweb** [comment]
Ran into this again today. I essentially have 4 options:

1. Run with `noUncheckedIndexedAccess: true`: Basically unusable because TypeScript doesn't let you narrow away the `| undefined`. Especially obnoxious for arrays, because TypeScript has no "inferred minimum length".
2. Wrap every `Record` with an index signature in `Partial<>`: Annoying because it basically just adds `| undefined` to the value type, which means you have to check for it in common operations like `for (const [key, value] of Object.entries(myRecord))` where you know that if a property exists, it's non-nullable (but there's nowhere to add the `!` operator to assert this).
3. Add an eslint-disable for every false positive.
4. Disable the rule.

None of those are particularly great options, so I hope the pull request gets some traction.

To be clear, I would actually like it if properties obtained via indexed access were treated as optional in various contexts and required me to check for that, but the available options for that (1 and 2) aren't great, so I don't think this rule should be reliant on them.

**Comment by ehoogeveen-medweb** [comment]
It looks like this was partially fixed by https://github.com/typescript-eslint/typescript-eslint/pull/8249.

The following example still produces a warning:
```ts
const foo: Record<string, [/* a truthy value type */]> = {};
const bar: string = 'not contained';
if (!foo[bar]) console.log('no such property');
```

[Playground](https://typescript-eslint.io/play#ts=5.3.3&fileType=.tsx&code=MYewdgzgLgBAZiEAuGAlApqATgEwDzRYCWYA5gDQwDaA9AFQwCGMUWArlABYCeMAbowA2bdC24AHUXRoBdAHwwAvDADeAXwDcAWABQoSLABGjLCkIlSSmAHIwIWPqiMS6HNe06icGAAoAhAggVMZYMgCUMPoQIILoAHSCIKQ%2BtiAwEGzAnDDiWCCSWFDc1mEaQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDArugDTg2ToAPAIbNcBYkNQAFTLERY8ATwAqi%2BWUod0vAL4gdQA&tokens=false)
